From 3d3341a1246b0fec96a559ba4ebbcf187e17f072 Mon Sep 17 00:00:00 2001
From: chandavath <rcrajagopal@habana.ai>
Date: Mon, 15 Jul 2024 19:57:33 +0300
Subject: [PATCH] [SW-185073] gaudi3: sec: dyncamically open sram region to load fit

This patch opens sram region dynamically to load the fit from host
(sudo hl-fw-loader).

Change-Id: I464ad9e2cf3087d70cb271c917a659b752081173
Signed-off-by: chandavath <rcrajagopal@habana.ai>
---

diff --git a/arch/gaudi3/common/security.c b/arch/gaudi3/common/security.c
index 14f8e65..5168a54 100644
--- a/arch/gaudi3/common/security.c
+++ b/arch/gaudi3/common/security.c
@@ -39,6 +39,9 @@
 #include <fit.h>
 #include <fit_rsa.h>
 #include <sha.h>
+#include <memory.h>
+
+#define ALIGN(x, a)		(((x) + (a) - 1) & ~((a) - 1))
 
 #define RR_HBW_PRIV_RANGE_MIN		0x0
 /* Below value is to keep HBW RRs configured at 16KB granularity */
@@ -1027,6 +1030,90 @@
 	}
 }
 
+static void hl_sec_config_dyn_hbw_rr(int rr, uint64_t min64, uint64_t max64)
+{
+	struct hl_security_rr_engine_select rr_sel;
+	struct hl_security_drv_data data;
+
+	get_security_info();
+	data.sku = !is_dual_die();
+	data.sec_dbg = is_security_debug_disabled();
+	data.hbw_rr_dbg_offset = RR_HBW_DBG_ARR_OFFSET;
+
+	memset(&rr_sel, 0, sizeof(struct hl_security_rr_engine_select));
+	/* Block pcie RR for dyn rr */
+	rr_sel.pcie = HL_SECURITY_PCIE_MSTR_RR;
+
+	hl_secure_hbw_rr_common(&data, rr, min64, max64, rr_sel);
+}
+
+#define IMAGE_OFFSET_MASK	0xFFFFFF
+void preboot_security_dynamic_img_sram_open(enum mem_region img_region,
+				__unused enum preboot_mem_access_mode mode)
+{
+	struct mem_desc mdesc;
+	uint64_t img_max, image_base_full;
+
+	if (!preboot_security_is_enabled())
+		return;
+
+	switch (img_region) {
+	case MEM_FWUPD_PCI:
+		if (preboot_get_mem_region(img_region, &mdesc))
+			preboot_puts_hex("SEC: failed requesting mem region ",
+					 img_region, false);
+		break;
+	default:
+		preboot_puts_hex("SEC: forbidden opening mem region ",
+				 img_region, false);
+		return;
+	}
+
+	image_base_full = (SRAM_BASE_ADDR + SRAM_MODE_0_SINGLE_DIE_OFFSET)
+						+ (mdesc.base & IMAGE_OFFSET_MASK);
+
+	/*
+	 * img_max is aligned to 16K to open the ranges according to image size
+	 * so complete image size range is opened
+	 */
+	img_max = ALIGN((image_base_full + mdesc.size), SZ_16K);
+
+	/*
+	 * Initially, when erom is booted with security enabled, the whole SRAM region is closed.
+	 * Let's assume, load fit (Size of 700Kb) in this SRAM region
+	 * (0x1D200000 - 0x1D2AF000). If we close this (RR_HBW_PRIV_START_HBW - 0x1D200000)
+	 * region first, then for a moment we are opening a region from
+	 * (0x1D200000 - RR_HBW_PRIV_START_RSVD), which will cause a problem.
+	 * So, first close (0x1D2AF000 - RR_HBW_PRIV_END_RSVD) and close
+	 * (RR_HBW_PRIV_START_HBW - 0x1D200000), which will open only the required region.
+	 */
+	hl_sec_config_dyn_hbw_rr(HL_SEC_HBW_RR_7, img_max, RR_HBW_PRIV_END_RSVD);
+
+	/*
+	 * In gaudi3 when RR are programmed max address is included.
+	 * Hence to open proper range for image size, max should be 16K page
+	 * earlier that image base address (HBW are configured in 16K pages aligned)
+	 */
+	hl_sec_config_dyn_hbw_rr(HL_SEC_HBW_RR_0, RR_HBW_PRIV_START_HBW,
+				(image_base_full - SZ_16K));
+}
+
+void preboot_security_dynamic_img_sram_close(void)
+{
+	uint64_t min, max;
+
+	if (!preboot_security_is_enabled())
+		return;
+
+	min = RR_HBW_PRIV_START_HBW;
+	max = RR_HBW_PRIV_END_HBW;
+	hl_sec_config_dyn_hbw_rr(HL_SEC_HBW_RR_0, min, max);
+
+	min = RR_HBW_PRIV_START_RSVD;
+	max = RR_HBW_PRIV_END_RSVD;
+	hl_sec_config_dyn_hbw_rr(HL_SEC_HBW_RR_7, min, max);
+}
+
 void preboot_security_boot_info_write_key_idx(u32 *key_idx)
 {
 	if (hl_boot_info_set(HL_BOOT_INFO_KEY_IDX, key_idx, sizeof(u32)))
